torch/csrc/autograd/variable.cpp

```diff
@@ -149,83 +154,88 @@ void Variable::Impl::backward(
 
 void Variable::Impl::set_data(Tensor new_data) {
   // Resets gradient accumulator if metadata is out of date
-  std::lock_guard<std::mutex> lock(mutex_);
-  auto prior_accumulator = grad_accumulator_.lock();
+  auto autograd_meta = get_autograd_meta();
+  std::lock_guard<std::mutex> lock(autograd_meta->mutex_);
+  auto prior_accumulator = autograd_meta->grad_accumulator_.lock();
   if (prior_accumulator) {
     const auto prior_device = prior_accumulator->input_metadata(0).device();
     const auto new_device = new_data.is_cuda() ? new_data.get_device() : -1;
 
     if (new_data.type() != data_.type() || prior_device != new_device) {
-      grad_accumulator_.reset();
+      autograd_meta->grad_accumulator_.reset();
     }
   }
 
   // Updates metadata
   data_type_ = new_data.type().typeMeta();
   type_id_ = new_data.type().type_id();
   is_variable_ = true;
-  data_ = std::move(new_data);
+
+  auto new_data_copy = at::Tensor(new_data.getIntrusivePtr()->shallow_copy_and_detach());
```

Comment:
why?

Reply:
Similar to the reasoning in another comment, the shallow copying is not absolutely necessary in this PR, but it is needed in the end state of the Variable/Tensor merge and would be good if we can have it in this PR (to avoid opening another PR just for it).